upbc/protoc-gen-upb.cc

```diff
@@ -369,6 +329,346 @@ std::string CTypeConst(const protobuf::FieldDescriptor* field) {
+  if (mt_f->submsg_index == kUpb_NoSub) {
+    return PackSub(nullptr, SubTag::kNull);
+  } else if (f->message_type()) {
+    return PackSub(AllocStr(MessageInit(f->message_type())), SubTag::kMessage);
+  } else {
+    ABSL_ASSERT(f->enum_type());
+    return PackSub(AllocStr(EnumInit(f->enum_type())), SubTag::kEnum);
+  }
+}
+
+const char* FilePlatformLayout::AllocStr(const std::string& str) {
+  char* ret =
+      static_cast<char*>(upb_Arena_Malloc(arena_.ptr(), str.size() + 1));
+  memcpy(ret, str.data(), str.size());
+  ret[str.size()] = '\0';
+  return ret;
+}
+
+void FilePlatformLayout::BuildMiniTables(const protobuf::FileDescriptor* fd) {
+  for (const auto& m : SortedMessages(fd)) {
+    table_map_[m] = MakeMiniTable(m);
+  }
+  ResolveIntraFileReferences();
+  SetSubTableStrings();
+}
+
+void FilePlatformLayout::BuildExtensions(const protobuf::FileDescriptor* fd) {
+  auto sorted = SortedExtensions(fd);
+  upb::Status status;
+  for (const auto& f : sorted) {
+    upb::MtDataEncoder e;
+    e.StartMessage(0);
+    e.PutField(static_cast<upb_FieldType>(f->type()), f->number(),
+               GetFieldModifiers(f));
+    upb_MiniTable_Extension& ext = extension_map_[f];
+    upb_MiniTable_Sub sub;
+    bool ok = upb_MiniTable_BuildExtension(e.data().data(), e.data().size(),
+                                           &ext, sub, status.ptr());
+    if (!ok) {
+      fprintf(stderr, "Error building mini-table: %s\n",
+              status.error_message());
+    }
+    ext.extendee = reinterpret_cast<const upb_MiniTable*>(
+        AllocStr(MessageInit(f->containing_type())));
+    ext.sub = PackSubForField(f, &ext.field);
+    ABSL_ASSERT(ok);
+  }
+}
+
+upb_MiniTable* FilePlatformLayout::MakeMiniTable(
+    const protobuf::Descriptor* m) {
+  if (m->options().message_set_wire_format()) {
+    return upb_MiniTable_BuildMessageSet(platform_, arena_.ptr());
+  } else if (m->options().map_entry()) {
+    return upb_MiniTable_BuildMapEntry(
+        static_cast<upb_FieldType>(m->map_key()->type()),
+        static_cast<upb_FieldType>(m->map_value()->type()),
+        m->map_value()->enum_type() &&
+            m->map_value()->enum_type()->file()->syntax() ==
+                protobuf::FileDescriptor::SYNTAX_PROTO3,
+        platform_, arena_.ptr());
+  } else {
+    return MakeRegularMiniTable(m);
+  }
+}
+
+upb_MiniTable* FilePlatformLayout::MakeRegularMiniTable(
+    const protobuf::Descriptor* m) {
+  upb::MtDataEncoder e;
+  e.StartMessage(GetMessageModifiers(m));
+  for (const auto& f : FieldNumberOrder(m)) {
+    e.PutField(static_cast<upb_FieldType>(f->type()), f->number(),
+               GetFieldModifiers(f));
+  }
+  for (int i = 0; i < m->real_oneof_decl_count(); i++) {
+    const protobuf::OneofDescriptor* oneof = m->oneof_decl(i);
+    e.StartOneof();
+    for (int j = 0; j < oneof->field_count(); j++) {
+      const protobuf::FieldDescriptor* f = oneof->field(j);
+      e.PutOneofField(f->number());
+    }
```

Comment:
do you need an `e.EndOneof()` or something?

Reply:
Nope, it's not currently required. Oneofs have separators between them, but no termination.